// is/index.js
module.exports = {
    '0': require('is/0'),
    arrayLike: require('is/array-like'),
    array: require('is/array'),
    boolean: require('is/boolean'),
    defined: require('is/defined'),
    emptyArray: require('is/empty-array'),
    empty: require('is/empty'),
    equal: require('is/equal'),
    false: require('is/false'),
    falsey: require('is/falsey'),
    finite: require('is/finite'),
    function: require('is/function'),
    greaterThan: require('is/greater-than'),
    http: require('is/http'),
    infinite: require('is/infinite'),
    instance: require('is/instance'),
    integer: require('is/integer'),
    key: require('is/key'),
    match: require('is/match'),
    nan: require('is/nan'),
    nil: require('is/nil'),
    null: require('is/null'),
    number: require('is/number'),
    object: require('is/object'),
    of: require('is/of'),
    promise: require('is/promise'),
    regExp: require('is/reg-exp'),
    same: require('is/same'),
    strictlyEqual: require('is/strictly-equal'),
    string: require('is/string'),
    symbol: require('is/symbol'),
    thennable: require('is/thennable'),
    true: require('is/true'),
    truthy: require('is/truthy'),
    undefined: require('is/undefined'),
    validInteger: require('is/valid-integer'),
    value: require('is/value'),
    window: require('is/window')
};
