// function/extend/index.js
var isString = require('is/string');
var merge = require('object/merge');
var has = require('object/has');
var isInstance = require('is/instance');
var factory = require('function/factory');
var bind = require('function/bind');
var PROTOTYPE = 'prototype';
var CONSTRUCTOR = 'constructor';
var FCC = Function[CONSTRUCTOR];
var EXTEND = 'extend';
var isValue = require('is/value');
var DOUBLE_UNDERSCORE = '__';
var COLON = ':';
var toArray = require('to/array');
var isOf = require('is/of');
var bindTo = require('function/bind-to');
var CONSTRUCTOR_KEY = DOUBLE_UNDERSCORE + CONSTRUCTOR + DOUBLE_UNDERSCORE;
var createFrom = require('object/create-from');
constructorExtend.wrapper = constructorWrapper;
module.exports = constructorExtend;

function constructorWrapper(Constructor, notOriginal) {
    __.isInstance = Constructor.isInstance = function (instance) {
        return isInstance(instance, Constructor);
    };
    __.factory = Constructor.factory = factory;
    __.fn = Constructor.fn = Constructor[PROTOTYPE].fn = Constructor[PROTOTYPE];
    __.constructor = Constructor.constructor = Constructor;
    __[EXTEND] = Constructor[EXTEND] = bind(constructorExtend, Constructor);
    __.origin = Constructor.origin = !notOriginal;
    return __;

    function __(one, two, three, four, five, six) {
        return isValue(one) && isOf(one, Constructor) ? one : new Constructor(one, two, three, four, five, six);
    }
}

function constructorExtend(name_, protoProps) {
    var options, nameString, constructorKeyName, child, passedParent, hasConstructor, constructor, parent = this,
        name = name_,
        origin = parent.origin,
        nameIsStr = isString(name);
    if (!nameIsStr) {
        protoProps = name;
        name = null;
    }
    hasConstructor = has(protoProps, CONSTRUCTOR);
    if (protoProps && hasConstructor) {
        child = protoProps[CONSTRUCTOR];
    }
    constructorString = 'return function ' + (name || this.name) + '(){\n\tvar args = [_.bindTo(_.supr, this)].concat(_.toArray(arguments));\n\treturn _.constructor.apply(this, args);\n}';
    // if (nameIsStr) {
    //     child = fcc(constructorString, this, parent);
    // } else {
    // if (child) {
    //     debugger;
    // }
    child = child ? fcc(constructorString, child, parent || this) : this;
    // }
    // child[EXTEND] = constructorExtend;
    var Surrogate = function () {
        this[CONSTRUCTOR] = child;
    };
    Surrogate[PROTOTYPE] = parent[PROTOTYPE];
    child[PROTOTYPE] = merge(createFrom(Surrogate), protoProps);
    // don't call the function if nothing exists
    constructor = child;
    child = constructorWrapper(constructor, 1);
    constructor[PROTOTYPE][CONSTRUCTOR_KEY] = child;
    return child;
}

function fcc(string, child, parent) {
    return new FCC('_', string)({
        toArray: toArray,
        bindTo: bindTo,
        supr: parent,
        constructor: child
    });
}
