// time/constants/index.js
var SIXTY = 60,
    SEVEN = 7,
    THIRTY = 30,
    TWENTY_FOUR = 24,
    ONE_THOUSAND = 1000,
    THREE_HUNDRED_SIXTY_FIVE = 365,
    SIXTY_THOUSAND = ONE_THOUSAND * SIXTY,
    THREE_HUNDRED_SIXTY_THOUSAND = SIXTY_THOUSAND * SIXTY,
    EIGHTY_SIX_MILLION_FOUR_HUNDRED_THOUSAND = THREE_HUNDRED_SIXTY_THOUSAND * TWENTY_FOUR,
    SIX_HUNDRED_FOUR_MILLION_EIGHT_HUNDRED_THOUSAND = THREE_HUNDRED_SIXTY_THOUSAND * SEVEN,
    TWO_BILLION_FIVE_HUNDRED_NINETY_TWO_MILLION = THREE_HUNDRED_SIXTY_THOUSAND * THIRTY,
    THIRTY_ONE_BILLION_FIVE_HUNDRED_THIRTY_SIX_MILLION = EIGHTY_SIX_MILLION_FOUR_HUNDRED_THOUSAND * THREE_HUNDRED_SIXTY_FIVE;
module.exports = {
    ms: 1,
    s: ONE_THOUSAND,
    secs: ONE_THOUSAND,
    seconds: ONE_THOUSAND,
    mins: SIXTY_THOUSAND,
    minutes: SIXTY_THOUSAND,
    hrs: THREE_HUNDRED_SIXTY_THOUSAND,
    hours: THREE_HUNDRED_SIXTY_THOUSAND,
    days: EIGHTY_SIX_MILLION_FOUR_HUNDRED_THOUSAND,
    wks: SIX_HUNDRED_FOUR_MILLION_EIGHT_HUNDRED_THOUSAND,
    weeks: SIX_HUNDRED_FOUR_MILLION_EIGHT_HUNDRED_THOUSAND,
    mnths: TWO_BILLION_FIVE_HUNDRED_NINETY_TWO_MILLION,
    months: TWO_BILLION_FIVE_HUNDRED_NINETY_TWO_MILLION,
    yrs: THIRTY_ONE_BILLION_FIVE_HUNDRED_THIRTY_SIX_MILLION,
    years: THIRTY_ONE_BILLION_FIVE_HUNDRED_THIRTY_SIX_MILLION
};
